### PR TITLE
Irrigation fields updated for the topounit-based subgrid structure

### PR DESCRIPTION
**Pull request by tktesfa2014**
This PR updates data structures of irrigation fields based on requirements of the topography-based subgrid structure. 

[BFB]

### CODE REVIEW

**Comment by tktesfa2014** [comment]
@hydrotian and @bishtgautam could you please review the changes related to the irrigation fields? Thanks.

**Review by bishtgautam — COMMENTED** [review]
@tktesfa2014 I have minor comments

**Comment by rljacob** [comment]
Is it BFB ?

File: components/elm/src/biogeophys/CanopyFluxesMod.F90 [hunk 1]
```diff
@@ -641,7 +646,7 @@ subroutine CanopyFluxes(bounds,  num_nolakeurbanp, filter_nolakeurbanp, &
                   ! Translate vol_liq_so and eff_porosity into h2osoi_liq_so and h2osoi_liq_sat and calculate deficit
                   h2osoi_liq_so  = vol_liq_so * denh2o * col_pp%dz(c,j)
                   h2osoi_liq_sat = eff_porosity(c,j) * denh2o * col_pp%dz(c,j)
-                  deficit        = max((h2osoi_liq_so + ldomain%firrig(g)*(h2osoi_liq_sat - h2osoi_liq_so)) - h2osoi_liq(c,j), 0._r8)
+                  deficit        = max((h2osoi_liq_so + firrig(g,tpu_ind)*(h2osoi_liq_sat - h2osoi_liq_so)) - h2osoi_liq(c,j), 0._r8)
```
Comment (hydrotian) [inline components/elm/src/biogeophys/CanopyFluxesMod.F90:649]: Will 'firrig' be the same in different topo-units for each grid? Same question to the other two variables 'f_grd' and 'f_surf'.

File: components/elm/src/main/surfrdMod.F90 [hunk 2]
```diff
@@ -886,7 +873,22 @@ subroutine surfrd_special(begg, endg, ncid, ns,ntpu)
          end do
       
     end do
+    
+    ! Obtain firrig and surface/grnd irrigation fraction
+    if (firrig_data) then
```
Comment (hydrotian) [inline components/elm/src/main/surfrdMod.F90:878]: The "firrig_data" was a flag indicating whether the surface data includes firrig, f_surf, and f_grd information. If not, then the default values (0.7, 1,0) will be used, so the user can still run the irrigation scheme. I would suggest keeping [lines 176-178](https://github.com/E3SM-Project/E3SM/pull/5016/files#:~:text=domain%25firrig%20%20%20%3D,%25f_grd%20%20%20%20%3D%200.0_r8) with the updated tpu indexing.

**Review by hydrotian — COMMENTED** [review]
Assuming that the three fields (firrig, f_surf, and f_grnd) are consistent over the TPUs for each grid. I'm not sure why we have to add the extra TPU dimension for these fields. Can we treat them like other domain fields like `area` or `frac`? Is there a way to check the dimension before reading the fields, if there's no TPU dimension?

**Comment by rljacob** [comment]
@tktesfa2014 please look at reviewer comments.

File: components/elm/src/biogeophys/BalanceCheckMod.F90 [hunk 3]
```diff
@@ -170,7 +170,7 @@ subroutine ColWaterBalanceCheck( bounds, num_do_smb_c, filter_do_smb_c, &
      use landunit_varcon   , only : istice_mec, istdlak, istsoil,istcrop,istwet
      use elm_varctl        , only : create_glacier_mec_landunit
      use elm_initializeMod , only : surfalb_vars
-     use domainMod         , only : ldomain
+     !use domainMod         , only : ldomain
```
[the remark this answers is not in the document]
Comment (tktesfa2014) [reply components/elm/src/biogeophys/BalanceCheckMod.F90:173]: Okay, I have addressed this.

File: components/elm/src/biogeophys/BalanceCheckMod.F90 [hunk 4]
```diff
@@ -383,7 +383,7 @@ subroutine ColWaterBalanceCheck( bounds, num_do_smb_c, filter_do_smb_c, &
              write(iulog,*)'qflx_evap_tot              = ',qflx_evap_tot(indexc)
              write(iulog,*)'qflx_irrig                 = ',qflx_irrig(indexc)
              write(iulog,*)'qflx_supply                = ',atm2lnd_vars%supply_grc(g)
-             write(iulog,*)'f_grd                      = ',ldomain%f_grd(g)
+             !write(iulog,*)'f_grd                      = ',ldomain%f_grd(g)
```
[the remark this answers is not in the document]
Comment (tktesfa2014) [reply components/elm/src/biogeophys/BalanceCheckMod.F90:386]: Okay, I have addressed this.

File: components/elm/src/biogeophys/BalanceCheckMod.F90 [hunk 5]
```diff
@@ -408,7 +408,7 @@ subroutine ColWaterBalanceCheck( bounds, num_do_smb_c, filter_do_smb_c, &
              write(iulog,*)'qflx_surf_irrig_col        = ',qflx_surf_irrig_col(indexc)
              write(iulog,*)'qflx_over_supply_col       = ',qflx_over_supply_col(indexc)
              write(iulog,*)'qflx_supply                = ',atm2lnd_vars%supply_grc(g)
-             write(iulog,*)'f_grd                      = ',ldomain%f_grd(g)
+             !write(iulog,*)'f_grd                      = ',ldomain%f_grd(g)
```
[the remark this answers is not in the document]
Comment (tktesfa2014) [reply components/elm/src/biogeophys/BalanceCheckMod.F90:411]: Okay, I have addressed this.

File: components/elm/src/biogeophys/CanopyFluxesMod.F90 [hunk 6]
```diff
@@ -100,7 +100,9 @@ subroutine CanopyFluxes(bounds,  num_nolakeurbanp, filter_nolakeurbanp, &
     use elm_varcon         , only : c14ratio
 
     !NEW
-    use domainMod          , only : ldomain
+    !use domainMod          , only : ldomain
```
[the remark this answers is not in the document]
Comment (tktesfa2014) [reply components/elm/src/biogeophys/CanopyFluxesMod.F90:103]: Okay, I have addressed this.

File: components/elm/src/biogeophys/CanopyFluxesMod.F90 (hunk 1 above)
Comment (tktesfa2014) [reply components/elm/src/biogeophys/CanopyFluxesMod.F90:649]: @hydrotian 

> Is it BFB ?

@rljacob yes it BFB. I just run SMS test and all passed.

Comment (tktesfa2014) [reply components/elm/src/biogeophys/CanopyFluxesMod.F90:649]: @hydrotian the implementation of the topographic units has a flag to check whether the surface data has topounits or not. If the surface data file has no topounits the code assigns a single topounit per grid. So, the changes I have implemented for the firrig, f_surf, and f_grnd are able to read both surface data files.

Comment (tktesfa2014) [reply components/elm/src/biogeophys/CanopyFluxesMod.F90:649]: @hydrotian the implementation allows both options: topounits of the same grid sharing the same value and topounits of the same grid assigned different values.

**Comment by tktesfa2014** [comment]
@bishtgautam yes this BFB. I have run SMS test and all passed.

**Comment by tktesfa2014** [comment]
@rljacob In the recent commit, I have addressed comments from @bishtgautam and @hydrotian . Thank you.

**Review by bishtgautam — APPROVED** [review]

File: components/elm/src/biogeophys/CanopyFluxesMod.F90 (hunk 1 above)
Comment (hydrotian) [reply components/elm/src/biogeophys/CanopyFluxesMod.F90:648]: Thanks for the explainations.

**Review by hydrotian — APPROVED** [review]

**Comment by tktesfa2014** [comment]
Thanks @bishtgautam and @hydrotian! @rljacob both reviewers have approved the PR. Thanks!

**Comment by rljacob** [comment]
@tktesfa2014 thanks.  I think this is ready then.

**Comment by tktesfa2014** [comment]
@rljacob please let me know if there is anything I need to do for this to be merged to master. Thanks.

**Comment by rljacob** [comment]
It is up to the integrator, @bishtgautam

**Pull request merged**